iree/turbine/runtime/op_reg/base.py

```diff
@@ -734,7 +769,37 @@ def mlir_type_asm(self) -> list[str]:
         return asms
 
 
-ArgDescriptor = Union[AttrArg, IntArg, TensorArg, TensorListArg]
+class EmptyOptionalTensorArg:
+    """Sentinel type marking an optional tensor argument that was not provided
+    at the call site.
+
+    To `KernelSelection` a `None` `ArgDescriptor` indicates an argument has been
+    declared as part of the signature, but the `ArgDescriptor` hasn't been
+    initialized with values an actual call site. An `EmptyOptionalTensorArg`
+    signals that an `ArgDescriptor` has been initialized, but an argument was
+    not provided at the call site.
+    """
+
+    ir_arity: int = 0
+    maybe_tensor_value: Optional[Tensor] = None
+    is_list: bool = False
+
+    def __repr__(self):
+        return "TensorArg(None)"
+
+    @property
+    def spec_key(self) -> str:
+        raise AssertionError("EmptyOptionalTensorArg has no spec_key")
+
+    @property
+    def mlir_type_asm(self) -> str:
+        raise AssertionError("EmptyOptionalTensorArg has no mlir_type_asm")
+
+    def generate_meta(self) -> list[Tensor]:
+        raise AssertionError("EmptyOptionalTensorArg has no meta generation")
```

Comment:
Now that I'm looking more carefully here, does it make sense to do the following? Not sure if this is ever going to be useful regardless, but at least the typing could be modified to not return a `list[Tensor]`. 
```suggestion
    def generate_meta(self) -> None:
        return None
```

Reply:
I think that's technically more correct. IIUC the only time this would matter would be if someone is trying to trace a model using a custom op that returns `None` value a custom op received as an output... I suppose `None` is the right return in that case 🤷‍♂️ .